jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java

```diff
@@ -526,34 +512,48 @@ public String toString()
                 inUse,
                 entries > 0 ? (inUse * 100) / entries : 0);
         }
+
+        private class BucketPool extends ConcurrentPool<RetainableByteBuffer>
+        {
+            private final boolean _direct;
+
+            public BucketPool(int poolSize, boolean direct)
+            {
+                super(StrategyType.THREAD_ID, poolSize, e -> 1);
+                _direct = direct;
+            }
+
+            @Override
+            protected void leaked()
+            {
+                (_direct ? ArrayByteBufferPool.this._currentDirectMemory : ArrayByteBufferPool.this._currentHeapMemory).addAndGet(-1 * _capacity);
```

Comment:
This solves the problem for leaked buffers originating from the `ConcurrentPool`, but not for leaked buffers coming from the `QueuedPool`.

Reply:
I think these atomic counters should go as keeping them in sync with the actual pooled buffer sizes could become overly complicated; we should instead make sure we don't need to calculate the pooled memory size too often.

BTW, calculating the total size isn't horribly expensive either: a bucket's memory size is just pool size * buffer size, an we need to iterate the bucket array to sum that all up.

Reply:
> This solves the problem for leaked buffers originating from the `ConcurrentPool`, but not for leaked buffers coming from the `QueuedPool`.

Can we leak from the queued pool?

Reply:
> I think these atomic counters should go as keeping them in sync with the actual pooled buffer sizes could become overly complicated; we should instead make sure we don't need to calculate the pooled memory size too often.
> 
> BTW, calculating the total size isn't horribly expensive either: a bucket's memory size is just pool size * buffer size, an we need to iterate the bucket array to sum that all up.

I'll give this a go. It should be rare and not too expensive.

Reply:
> Can we leak from the queued pool?

Sure, the queued pool is naturally leak-resistant, but any code using it may still leak a buffer.

What I mean is that when the above occurs, there is no leak detection feature in the queued pool that we can plug into to update the atomic sizes.